File: src/editor/database/card/SplitCard.java
package editor.database.card;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import editor.util.Lazy;

/**
 * This class represents a card that has several faces all printed on the front.
 *
 * @author Alec Roelke
 */
public class SplitCard extends MultiCard
{
    /**
     * Converted mana costs of this SplitCard's faces.
     */
    private Lazy<List<Double>> cmc;

    /**
     * Create a new SplitCard with the given Cards as faces.
     *
     * @param f cards to use as faces
     */
    public SplitCard(Card... f)
    {
        this(Arrays.asList(f));
    }

    /**
     * Create a new SplitCard with the given Cards as faces.  They should all indicate
     * with their layouts that they are split cards.
     *
     * @param f list of Cards representing faces
     */
    public SplitCard(List<Card> f)
    {
        super(f.get(0).layout(), f);
        for (Card face : f)
            if (face.layout() != CardLayout.SPLIT && face.layout() != CardLayout.AFTERMATH)
                throw new IllegalArgumentException("can't create split cards out of non-split cards");
        for (Card face : f)
            if (face.layout() != f.get(0).layout())
                throw new IllegalArgumentException("all faces of a split card must be of the same type");

        cmc = new Lazy<>(() -> Collections.unmodifiableList(Collections.nCopies(f.size(), f.stream().mapToDouble((c) -> c.cmc().get(0)).sum())));
    }

    /**
     * {@inheritDoc}
     * The converted mana cost of a split card is the sum of the converted mana costs of all
     * of its faces.
     */
    @Override
    public List<Double> cmc()
    {
        return cmc.get();
    }

    /**
     * {@inheritDoc}
     * All of the faces of a SplitCard are on the front, so there is only one image for it.
     */
    @Override
    public List<String> imageNames()
    {
        return super.imageNames().subList(0, 1);
    }

    /**
     * {@inheritDoc}
     * All of the faces of a SplitCard are on the front, so only one multiverseid is necessary.
     */
    @Override
    public List<Long> multiverseid()
    {
        return super.multiverseid().subList(0, 1);
    }
}

File: src/main/scala/editor/database/card/SplitCard.scala
package editor.database.card

import CardLayout.{ADVENTURE, AFTERMATH, SPLIT}

/**
 * A [[Card]] with all faces printed on the front, with any one castable at a time (with some exceptions).
 * The mana value of such a card is the sum of its faces.
 * 
 * @constructor create a new split card
 * @param faces individual mini cards on the front of the split card
 * @author Alec Roelke
 */
@throws[IllegalArgumentException]("if any face isn't a type of split card or all faces are not the same type")
class SplitCard(faces: Seq[Card]) extends MultiCard(faces(0).layout, faces.toIndexedSeq) {
  if (!faces.forall((f) => Seq(SPLIT, AFTERMATH, ADVENTURE).contains(f.layout)))
    throw IllegalArgumentException("can't create split cards out of non-split cards")
  if (!faces.forall(_.layout == faces(0).layout))
    throw IllegalArgumentException("all faces of a split card must be of the same type")
  
  override lazy val manaValue = layout match {
    case SPLIT | AFTERMATH => faces.map(_.manaValue).sum
    case ADVENTURE => faces(0).manaValue
    case _ => throw IllegalStateException(s"$layout isn't a split layout")
  }
  override def minManaValue = faces.map(_.manaValue).min
  override def maxManaValue = faces.map(_.manaValue).max
  override def avgManaValue = faces.map(_.manaValue).sum/faces.size
  override def imageNames = faces(0).imageNames
  override def multiverseid = faces(0).multiverseid
  override def scryfallid = faces(0).scryfallid
}